### PR TITLE
Allow underscores in provider namespace

### PR DESCRIPTION
**Pull request by lbordowitz**
Following @apparentlymart idea [in this comment](https://github.com/opentofu/registry-address/pull/8#issuecomment-4207972519). I've also added a couple of checks for prefix/suffix underscores, as well as consecutive underscores.

The tests are entirely written by @nakhthar, who should probably get most of the credit for this PR given the amount of work they put in triage, determining the area to change, etc. @nakhthar is the original author of #8 which this PR is based off of.

### CODE REVIEW

File: provider.go [hunk 1]
```diff
@@ -457,18 +457,32 @@ func ParseProviderPart(given string) (string, error) {
...
+	// prefix or suffix, but does allow underscores. We do not.
+	if strings.HasPrefix(given, "_") || strings.HasSuffix(given, "_") {
+		return "", fmt.Errorf("underscores may not be used as a prefix or suffix")
+	}
 
-	result, err := idna.Lookup.ToUnicode(given)
+	unicodeProfile := idna.New(
+		idna.MapForLookup(),
+		idna.BidiRule(),
+		idna.StrictDomainName(false),
```
Comment (apparentlymart) [inline provider.go:485]: This option is opting us out of rejecting characters that are not [allowed in domain name labels](https://github.com/golang/net/blob/6c6f714fe2e55253602c0286e1e4bddfd571c2b7/idna/idna.go#L768-L774), so we should probably now add some additional tests to cover some more negative cases that return errors... it looks like the original tests here were focused mainly on testing valid cases and assuming that the idna library's own tests would cover the main invalid stuff.

For example, I think we might no longer have anything checking that the name doesn't contain arbitrary ASCII punctuation characters like `&`, `|`, etc.

Perhaps we should create our own variant of that upstream `allowedSTD3` function that extends it to also include `r == '_'` and run it on the result of `ToUnicode` below just so we can make sure we're _only_ adding support for underscores and not inadvertently widening it to include all of the ASCII punctuation too. 🤔

Comment (lbordowitz) [reply provider.go:485]: That seems a little complicated. How about this?

1. Check the domain name with underscores using the method in this PR. Save the normalized output.
2. Remove all underscores and only underscores from the string, run a strict check using the `idna.Lookup` profile like before. Check for errors
3. If there are no errors, return the normalized output from before, rather than the `Lookup` output.

Something we might want to validate against manually is `-_` and `_-` strings, which are ugly but technically valid I guess. I don't like them and we should test against them.

I'll add some tests with more of the ASCII punctuation characters.

Comment (apparentlymart) [reply provider.go:485]: To be honest, doing two calls to `ToUnicode` as you've described feels _more_ complicated to me, but of course that's subjective.

My suggestion of making our own version of `allowedSTD3` was coming from the idea that we're intentionally moving away from following the UTS 46/STD 3 rules to our own custom rules, and so having _our_ rule represented by a self-contained function in our own codebase seems clearer to me: it means that a future maintainer can refer just to that function to understand which characters we permit, instead of having to think about what's allowed by a combination of both upstream and local checks.

I don't feel super strongly about it, though: the important thing is to make sure we'd not start accepting any input we wouldn't have accepted before (aside from the intentional change to now allow underscores), and then we can always change the implementation details of that later if we want to.

Comment (lbordowitz) [reply provider.go:485]: So, should I re-implement the `mapping` function? Because if we want to add a custom `Option` for a fresh profile, the function we provide as that option is either toggling booleans or setting functions [in the `options` struct](https://github.com/golang/net/blob/v0.40.0/idna/idna10.0.0.go#L177-L197). The `mapping` function used in lookup [is slightly complicated](https://github.com/golang/net/blob/v0.40.0/idna/idna10.0.0.go#L475), but it involves [looking up characters](https://github.com/golang/net/blob/v0.40.0/idna/tables15.0.0.go#L625) in a [big fat block](https://github.com/golang/net/blob/v0.40.0/idna/tables15.0.0.go#L727) that encodes all the relevant characters' categories. It'd be cool if there was a regex I could swap out or a more general option that allows for fine-tuning within the STD3 lookup/mapping. I don't think there is in this version of IDNA, unless I missed something.

I'm still trying to figure out how this code is generated so I can understand it a little better. ETA [found it](https://github.com/golang/text/blob/master/internal/export/idna/idna.go#L5)

Comment (apparentlymart) [reply provider.go:485]: Oh, I was meaning to just run our custom function _after_ we've called `ToUnicode` on the custom profile where we disabled the `allowedSTD3` call, to make sure that the result doesn't include any characters we don't want.

That is admittedly slightly different than what the built-in `StrictDomainName` check would do, since that works on a label-by-label basis instead of on the entire string. However, since we're rejecting `'.'` in an earlier check above, we can assume that our input and result will always have just one DNS label, so I think that checking the entire result should be functionally equivalent.

Comment (lbordowitz) [reply provider.go:485]: Reading [the standard](https://www.unicode.org/reports/tr46/#UseSTD3ASCIIRules) helped a lot; I was kinda stuck in the weeds with that code! Hopefully the fix I put in is reasonable and well-documented enough.

**Review by nakhthar — COMMENTED** [review]
@lbordowitz @apparentlymart  any update on this PR ?

**Comment by apparentlymart** [comment]
Looking back on the previous discussion I realize that the discussion about adding some additional tests for disallowed ASCII punctuation characters in https://github.com/opentofu/registry-address/pull/9#discussion_r3094480996 became a discussion about how to implement that rule instead, and the implementation of the rule changed but we didn't end up with those extra tests. 

@lbordowitz would you like to add those tests before we move forward here, or did you skip them intentionally?

**Comment by lbordowitz** [comment]
I think I would like to add those additional tests, especially since we're changing the lookup profile. Let me get right on that.

**Comment by lbordowitz** [comment]
I added a couple punctuation tests. If you think it ought to be exhaustive, I can write a test where it changes the middle character to ASCII decimal 32-47 (accept 45 `-`), 58-64, 91-96 (accept 95 `_`), 123-126.

**Review by apparentlymart — APPROVED** [review]
Given that our character validation rule is defined by what is allowed rather than by what is disallowed, I think just having this selection of example disallowed characters in our test cases is a fine compromise, just as a way to verify that the `validDomainASCII` check is being reached at all. Testing this more precisely would amount to duplicating the `validDomainASCII` implementation in the test case in a slightly different shape, and I think that'd have dubious benefit.

**Comment by lbordowitz** [comment]
Merge is blocked pending discussion of acceptance by maintainers. In our usual process, we create an issue, determine whether it's appropriate, assign someone to the issue, and then review the PR. It seems no issue was created, so we'll check back on this Tuesday of next week to determine if this is something we even want in the OpenTofu project.

**Comment by diofeher** [comment]
@lbordowitz 

1. Did you see the work needed to integrate these changes in other repos, like the registry submission and the core project as well?

2. One thing that I don't understand: is there any valid reason for using IDNA if we do not allow unicode characters and only accept a subset of ASCII characters? Wouldn't be easy to just remove this dependency instead of trying to circumventing it?

**Comment by apparentlymart** [comment]
The call to `validDomainASCII` happens only if the character is already in the ASCII range, so this still allows non-ASCII characters. This should just cause underscore to be allowed where it wasn't before, without denying anything that was previously valid.

The use of the IDNA library here is mainly just to give us access to its implementation of [Nameprep](https://en.wikipedia.org/wiki/Nameprep) case folding, so that non-ASCII letters are normalized to lowercase in the same way they would be in domain name labels. Here we've just taken more control over the validation step, while keeping the upstream case folding implementation.

**Comment by lbordowitz** [comment]
We are not going to be implementing this feature. See #10 for more details.
